### ISSUE TITLE
Sprite collision detection

### ISSUE THREAD

**Issue by dirkwhoffmann** [dirkwhoffmann/vAmiga#79]
Since dual-playfields are supported now, vAmiga is nearly feature complete. Woohoo 🍺! 
(Of course, there are trillions of bugs and compatibility issues to be worked on.) 

The only missing feature (If I don't miss anything) is sprite collision detection.

Does somebody know a decent test-case for this? Professional games are not well suited for testing, because only a few of them use collision detection at all.

**Comment by Alessandro1970** [comment]
Could you relase a beta of vAmiga to try it ?

**Comment by mithrendal** [comment]
Archon and gravityforce are using sprite to sprite and sprite to playfield collision detection. 

Most of emulators for android do not proper implement this. Three years ago I backported the code into the iAmiga project

Emufreak/iAmiga#19

**Comment by dirkwhoffmann** [comment]
The current version can be downloaded here:

http://www.dirkwhoffmann.de/vAmiga/vAmiga_0.26.zip

Don't be confused that the mouse pointer is now visible outside the main screen. Making shift register handling more accurate has broken the old border drawing code.

**Comment by dirkwhoffmann** [comment]
> Archon and gravityforce

That's good to know! Unfortunately, both games do not run yet. 🙁

Best would be some kind of a demo program (maybe on some Fish disk) that allows the user to move some simple sprites and to inspect the detection register.

**Comment by dirkwhoffmann** [comment]
Sprite collision detection is up and running.

Current scores (tests collision1, ..., collision7 from vAmigaTS repo): 

vAmiga: 7/7 😎
UAE: 6/7 🤭

OK, I have to admit that UAE's better than vAmiga in every other respect 😉.

Because collision detection is both expensive and rarely used, I added three user-options to the compatibility preferences letting the user enable or disable sprite-sprite, sprite-playfield, and 
playfield-playfield collision detection individually. UAE  and SAE have similar options.

**Comment by mithrendal** [comment]
<img width="778" alt="image" src="https://user-images.githubusercontent.com/17108995/62677501-d32e8480-b9ae-11e9-89c5-e02038d4a05b.png">

Seems as  sprite to playfield test  in Archon II fails ? The guy in the fire is supposed to loose all his energy. But he is invincible .... Although simple shooting collision works (sprite to sprite) playfield to sprite seems not.  

[disk.adf.zip](https://github.com/dirkwhoffmann/vAmiga/files/3479999/disk.adf.zip)

**Comment by mithrendal** [comment]
@dirkwhoffmann  I tested this with the latest changes from vAmiga/master this morning. Will do some cross check tests in SAE this evening...

**Comment by dirkwhoffmann** [comment]
Wasn't Archon the game that didn't load in previous releases? 

BTW, I am working on the ExtRom branch most of the time and only merge back into the master branch occasionally. Therefore, the master branch is not always up-to-date.

I'll give Archon a try and report back...

**Comment by dirkwhoffmann** [comment]
I'm really thrilled that both Archon and Archon II do run in vAmiga now 😀.

<img width="858" alt="Bildschirmfoto 2019-08-08 um 14 15 43" src="https://user-images.githubusercontent.com/12561945/62704048-cc256780-b9ea-11e9-88e9-1a5ce13a8c5e.png">

<img width="858" alt="Bildschirmfoto 2019-08-08 um 14 18 21" src="https://user-images.githubusercontent.com/12561945/62703478-a2b80c00-b9e9-11e9-9916-8b43f0362f96.png">

Sprite-to-sprite collisions seems to work correctly. In Archon I, if I let my Orc knock his stick over the knight's skull, I can see the following debug message in the console (if CLX_DEBUG is set to 1): 

    [5644] (122,226)  037E3C BCBSDA 602C 0440 Denise: Collision between 01 and 67

Also, the knight dies when I do that. (Poor guy, but it's all in the sake of science 🙄.) 

I didn't check sprite / playfield collisions, because I don't know how the real game behaves. I would be surprised if sprite / playfield collision don't work, because test cases collision1 to collision7 all pass in vAmiga now. There might well be a bug of course, because my collision tests do not cover all possible combinations though (there are just too many). 

The master branch is up-to-date now. I've just merged in the ExRom branch.

**Comment by mithrendal** [comment]
In Archon II you have the four playgrounds to fight in: earth water air and fire. In each you have a wizard who can spell casts for example summon an orc. Each playground has its specific battle-background -lements: earth has some rocks, fire some fire fields.. When you let your character on a earth battlegorund  run against a rock hen it can not walk through. It is stopped before the rock. When you let your character run into the fire on a fire battleground then it burns and its energy is depleted continuosly. 

What I suspect is, that these background graphics is sprite to playfield collision.

**Comment by dirkwhoffmann** [comment]
>  When you let your character run into the fire on a fire battleground then it burns and its energy is depleted continuosly.

Indeed, I can run into the fire without loosing energy. I'll investigate what's happening there...

**Comment by dirkwhoffmann** [comment]
What I found out so far: 

Left player is sprite 6
Right player is sprite 7

When I run into the fire (with sprite 7), Denise detects a collision with PF2: 

[7301] (102,226)  02C298 BCBSDA 602C 01C0 Denise: S6 collides with PF2
[7302] ( 99,226)  02C292 BCBSDA 602C 01C0 Denise: S6 collides with PF2

The debug message is misleading (S6 is sprite 7, because counting starts at 0 internally).

Time to check how the game behaves on a real machine 😎. Stay tuned...

**Comment by dirkwhoffmann** [comment]
More findings: 

When I walk into the fire, the following happens internally: 

    [8005] ( 99,226)  02C298 BCBSD- 602C 0040 Denise: <6> b[769] = 2A e1 = 10 e2 = 2A, c1 = 10 c2 = 2A
    [8005] ( 99,226)  02C298 BCBSD- 602C 0040 Denise: S6 collides with PF2
    [8005] ( 99,226)  02C298 BCBSD- 602C 0040 Denise: <6> b[755] = 2A e1 = 10 e2 = 2A, c1 = 10 c2 = 2A
    [8005] (100,226)  02C28E BCBSD- 602C 0040 Denise: S6 collides with PF2
    [8005] (100,226)  02C28E BCBSD- 602C 0040 Denise: <6> b[769] = 2A e1 = 10 e2 = 2A, c1 = 10 c2 = 2A
    [8005] (100,226)  02C28E BCBSD- 602C 0040 Denise: S6 collides with PF2
    [8005] (100,226)  02C28E BCBSD- 602C 0040 Denise: <6> b[755] = 2A e1 = 10 e2 = 2A, c1 = 10 c2 = 2A
    [8005] (101,226)  02C298 BCBSD- 602C 0040 Denise: S6 collides with PF2
    [8005] (101,226)  02C298 BCBSD- 602C 0040 Denise: <6> b[769] = 2A e1 = 10 e2 = 2A, c1 = 10 c2 = 2A
    [8006] ( 99,226)  02C298 BCBSD- 602C 0040 Denise: S6 collides with PF2

Assuming that the bitplane data is correct, Denise is doing the right thing. It detects a collision with PF2, but not with PF1.

I also observed the following in vAmiga:  

When I just hit a few pixels of the fire, no energy is lost (I guess these pixels are drawn in PF1): 

<img width="82" alt="Bildschirmfoto 2019-08-08 um 15 58 18" src="https://user-images.githubusercontent.com/12561945/62709619-00eaec00-b9f6-11e9-915e-6a95ecdb6573.png">

But when I do this, energy does get lost: 

<img width="82" alt="Bildschirmfoto 2019-08-08 um 15 58 25" src="https://user-images.githubusercontent.com/12561945/62709650-0e07db00-b9f6-11e9-814b-cfc33e55638a.png">

This means that you do loose energy in the fire (in vAmiga), but you have to walk in the middle of it. My hypothesis is that it is a game feature. Hitting just a few fire pixels are not supposed to hurt. 

To reproduce, play human vs. human to have enough time for experiments.

I need to doublecheck on my A500...

**Comment by dirkwhoffmann** [comment]
Just doublechecked...

There is indeed an issue. In vAmiga, it seems like a collision is detected only if you walk through the bottom part of the fire. If there are three fires in a row, you can just walk through the upper and middle fire without getting hurt.

On the A500, you can do the following without being hurt, but you can never walk through a fire as it is possible in vAmiga: 

<img width="257" alt="Bildschirmfoto 2019-08-08 um 16 33 40" src="https://user-images.githubusercontent.com/12561945/62712026-5923ed00-b9fa-11e9-87e1-d124160e2daa.png">

Bottom line: vAmiga misses some collisions.

**Comment by dirkwhoffmann** [comment]
Nailed it.  Now fire hurts in vAmiga, too 🤯. 

The bug was hidden in function Denise::drawSprites(): 

The culprit is this code portion: 

     if (col) {
            if (z > zBuffer[pos]) {
                iBuffer[pos] = baseCol | col;
                zBuffer[pos] |= z;
            }
            if (z > zBuffer[pos-1]) {
                iBuffer[pos-1] = baseCol | col;
                zBuffer[pos-1] |= z;
            }
        }

Background: vAmiga utilizes a z buffer to implement the display hierarchy. Then depth value is chosen in a way that it codes some meta-information besides the depth of the pixel. I.e., there is a dedicated bit for each sprite indicating if the sprite is solid at this pixel position. In the old code, the sprite flag was only set when the sprite was in front which is wrong (the flag has to be set whenever the sprite pixel is solid).

Here is the (easy) fix: 

      if (col) {
            if (z > zBuffer[pos]) iBuffer[pos] = baseCol | col;
            if (z > zBuffer[pos-1]) iBuffer[pos-1] = baseCol | col;
            zBuffer[pos] |= z;
            zBuffer[pos-1] |= z;
        }

The fix has already been checked in and merged into the main branch.

**Comment by mithrendal** [comment]
Just tested the firewalk. It works as expected now I would say ! Very cool. 

I wanted to test more sprite2playfield collisions in games... One nice candidate for sprite to playfield collisions is gravityforce, but it does not start at all yet. We can test this somewhen later when vAmiga has gotten more refined over the time....

Since Archon is a very nice test candidate for both sprite2pf and sprite2sprite collision. And it runs so well in vAmiga. Dirk thats really awesome.

Rotor is the only game I know out there which uses playfield to playfield collision detection. I will give it a try in the evening...

**Comment by mithrendal** [comment]
tried to check rotor for playfield2playfield ... but it get stuck while loading.... shortly before it enters what seems to be an endless loop ... there we can spot these debug output lines ... 

`
[634] (281,211)  00F014 BCB-D- 606C 1780 [0245E0] Copper: ***** BFD FLAG NOT IMPLEMENTED YET
[634] (288, 46)  00F13A BCB-D- 606C 1780 [0245E4] Copper: ***** BFD FLAG NOT IMPLEMENTED YET
[634] (288, 47)  00F13A BCB-D- 606C 1780 [0245E4] Copper: ***** BFD FLAG NOT IMPLEMENTED YET
[634] (288, 48)  00F13C BCB-D- 606C 1780 [0245E4] Copper: ***** BFD FLAG NOT IMPLEMENTED YET
[634] (302,  6)  00F0E0 BCB-D- 606C 1780 [0245E8] Copper: ***** BFD FLAG NOT IMPLEMENTED YET
[634] (303, 30)  00F0B8 BCB-D- 606C 1780 [0245EC] Copper: ***** BFD FLAG NOT IMPLEMENTED YET
`


maybe it is also too early for that game test ... we have to come back here somewhen later I guess...  🤓

**Comment by dirkwhoffmann** [comment]
The "infinite loop" seems to be created simply by the huge amount of warning messages of type

    Copper: ***** BFD FLAG NOT IMPLEMENTED YET

It's generated in Copper.cpp

    case COP_WAIT1:
            
            if (verbose) debug("COP_WAIT1\n");

            if (getBFD() == 0) {
                debug("***** BFD FLAG NOT IMPLEMENTED YET\n");
            }

            // Wait for the next free cycle
            if (!agnus->copperCanRun()) { reschedule(); break; }

If the debug message is disabled, the game seems to run just fine 😎.

<img width="858" alt="Bildschirmfoto 2019-08-10 um 07 11 09" src="https://user-images.githubusercontent.com/12561945/62817887-7235b680-bb3e-11e9-9a08-a1b9d84f32d0.png">

<img width="858" alt="Bildschirmfoto 2019-08-10 um 07 06 51" src="https://user-images.githubusercontent.com/12561945/62817888-75c93d80-bb3e-11e9-8b81-c33038261d89.png">

However, I'm not sure if PF/PF collision works, because this game is damn difficult 😖. I've broken 5 space ships in less than a minute 🤭. On the positive side, this means that the game does detect collisions between the ship and the wall.

**Comment by dirkwhoffmann** [comment]
I've just uploaded a new pre-release (0.32).

It should be compatible with Archon, Archon II, Sword of Sodan, and Rotor. 

This time, I've uploaded it directly on Github: 

<img width="591" alt="Bildschirmfoto 2019-08-10 um 10 08 23" src="https://user-images.githubusercontent.com/12561945/62819551-13c90200-bb57-11e9-8f23-96f184140b6f.png">

It also comes with a brand new reset button 😎: 

<img width="146" alt="Bildschirmfoto 2019-08-10 um 09 42 33" src="https://user-images.githubusercontent.com/12561945/62819593-9a7ddf00-bb57-11e9-9fc4-037e3979f566.png">

For implementing the reset, I had to shift a lot of stuff from functions _powerOn() to functions _reset(). When moving large portions of code, there is always the risk of breaking something. So you better fasten your seat-belts before pushing the button.

**Comment by mithrendal** [comment]
Playfield2Playfield Collision detection is working -> Confirmed

When you only disable pf/pf collision detection in the compatibility preferences of vAmiga then the spaceship flies through the walls (and the game goes totally crazy ... ) Your ship can even fly beyond the boundary which the programmer has expected to be normal it seems. If you fly always in one direction then you fly into the outerspace of the chipram ([Kuiper belt](https://en.wikipedia.org/wiki/Kuiper_belt) ?). Inevitably the game crashes then with a completely distorted random pixelated screen. I guess the game/blitter draws the spaceship everywhere in the memory ... maybe even in programm code which leads to a crash ... no MMU here ?

In contrast when pf/pf collision is enabled in vAmiga prefs, everything works as expected. 

Conclusion: vAmiga is fully capable and compatible with all types of amiga hardware collision detection ! And that is soooo nice !!!

**Comment by dirkwhoffmann** [comment]
> Your ship can even fly beyond the boundary which the programmer has expected to be normal it seems.

I suppose the spaceship first thought: "Oh cool, the walls are gone, I can escape". And then seconds later it fell over the Chip Ram cliff. That's a really sad fate. My own spaceships didn't do better. None of them reached level 2. 😢

> Conclusion: vAmiga is fully capable and compatible with all types of amiga hardware collision detection !

🥳 Woohoo. I'll close this issue then and hope that we'll never have to reopen it again.
